crates/evm/evm/src/executors/invariant/shrink.rs

```diff
@@ -97,56 +97,73 @@ pub(crate) fn shrink_sequence(
     let mut shrinker = CallSequenceShrinker::new(calls.len());
     for _ in 0..failed_case.shrink_run_limit {
         // Check candidate sequence result.
-        match check_sequence(failed_case, executor.clone(), calls, shrinker.current().collect()) {
+        match check_sequence(
+            executor.clone(),
+            calls,
+            shrinker.current().collect(),
+            failed_case.addr,
+            failed_case.func.clone(),
+            failed_case.fail_on_revert,
+        ) {
             // If candidate sequence still fails then shrink more if possible.
-            Ok(false) if !shrinker.simplify() => break,
+            Ok((false, _)) if !shrinker.simplify() => break,
             // If candidate sequence pass then restore last removed call and shrink other
             // calls if possible.
-            Ok(true) if !shrinker.complicate() => break,
+            Ok((true, _)) if !shrinker.complicate() => break,
             _ => {}
         }
     }
 
     Ok(shrinker.current().map(|idx| &calls[idx]).cloned().collect())
 }
 
-/// Checks if the shrinked sequence fails test, if it does then we can try simplifying more.
-fn check_sequence(
-    failed_case: &FailedInvariantCaseData,
+/// Checks if the given call sequence breaks the invariant.
+/// Used in shrinking phase for checking candidate sequences and in replay failures phase to test
+/// persisted failures.
+/// Returns the result of invariant check and if sequence was entirely applied.
+pub fn check_sequence(
     mut executor: Executor,
     calls: &[BasicTxDetails],
     sequence: Vec<usize>,
-) -> eyre::Result<bool> {
+    test_address: Address,
+    test_function: Bytes,
+    fail_on_revert: bool,
+) -> eyre::Result<(bool, bool)> {
     let mut sequence_failed = false;
-    // Apply the shrinked candidate sequence.
+    // Apply the call sequence.
+    // We keep track of the number of calls applied to check if persisted sequence is
+    // replayed entirely.
+    let mut calls_applied = 0;
     for call_index in sequence {
+        calls_applied += 1;
+
         let (sender, (addr, bytes)) = &calls[call_index];
         let call_result =
             executor.call_raw_committing(*sender, *addr, bytes.clone(), U256::ZERO)?;
-        if call_result.reverted && failed_case.fail_on_revert {
+        if call_result.reverted && fail_on_revert {
             // Candidate sequence fails test.
             // We don't have to apply remaining calls to check sequence.
             sequence_failed = true;
             break;
```

Comment:
can't we just return here instead of managing `sequence_failed` flag and counting calls for `replayed_entirely`?

Reply:
yep yep, nice cleanup! https://github.com/foundry-rs/foundry/pull/7899/commits/7686c03398e99f36af8462db6798c579e229827b